### ISSUE TITLE
Add SQL query `limit` parameter to the search functions

### ISSUE THREAD

**Issue by fundthmcalculus** [Dentity-Development/dentity-ecosystems-sdk#597]
Each language should already have a `Search(string sqlQuery)` function. Each language should also already have a `Search()` function with a sane SQL query default. This ticket is a request to add the `limit` parameter as an optional argument to the sane defaults `Search()` function so that the user can limit the count of results returned. Default remains to return all of them.

### Details of `limit` parameter
* Essentially just switches from `SELECT * FROM c` to `SELECT * FROM c LIMIT [N=10]` (where `N` can be changed, obviously)
* Do we want a default `order by` clause as well to ensure a stable query?

### Requirements
1. Hide `continuation_token` and `has_more` parameters, return everything from the user's query
2. Optional limit parameter on the sane-defaults query

### WalletService
Languages:
- [x] .NET
- [x] Go
- [x] Java
- [x] Node
- [x] Python
- [x] Ruby
- [x] Swift
- [x] Web

### TrustRegistryService
Languages:
- [x] .NET
- [x] Go
- [x] Java
- [x] Node
- [x] Python
- [x] Ruby
- [x] Swift
- [x] Web

**Comment by tmarkovski** [comment]
Can you add more details what this `limit` refers to? I'm not very clear. Is this the LIMIT in the SQL query?

**Comment by fundthmcalculus** [comment]
> Can you add more details what this `limit` refers to? I'm not very clear. Is this the LIMIT in the SQL query?

Updated, see above.

**Comment by sethjback** [comment]
> Can you add more details what this limit refers to? I'm not very clear. Is this the LIMIT in the SQL query?

My understanding is this would limit the results returned from the query.

This functionality is arguably already provided by allowing the user to craft their own sql query and include a "limit" clause. That being the case, it still seems more user friendly to provide a simple way for users to limit or paginate the default query rather than writing SQL. The idea is their first instinct is not going to be "write SQL" but "where are the pagination flags"

I know we started this discussion elsewhere, and to continue it now: is there a need for the limit flag, or is pagination a better option?

Thinking of scenarios: if there is a large number of returned items by default, and the user wants to filter, the best option might be to use the SQL functionality.

On the other hand, if there are a large number of results and the users wants them all, they don't necessarily want a limit, but the ability to paginate.

Thoughts?

**Comment by tmarkovski** [comment]
I think we're sort of requiring developers to use some kind of SQL for properly working with the result sets. We added the default query in the search as a convenience, but I think abstracting that further might be hard. The SQL allows full control over paged and sorted data. Abstracting this in a more convenient data model with page index, page size etc might be good, but it requires that we inject and parse their SQL code.
We need more thought on how to do this correctly.

**Comment by tmarkovski** [comment]
Here's how users can paginate currently

Page 1
`trinsic template search --query 'SELECT * FROM c OFFSET 0 LIMIT 5'`
Page 2
`trinsic template search --query 'SELECT * FROM c OFFSET 5 LIMIT 5'`
Page 3
`trinsic template search --query 'SELECT * FROM c OFFSET 10 LIMIT 5'`

Get total for their query
`trinsic template search --query 'SELECT COUNT(c.id) as total FROM c'`

**Comment by tmarkovski** [comment]
Another way to query would be not to use pages, but rely on the continuation token functionality, in which case CosmosDB will decide how many results it will return which is 1000 or 1MB result, whichever limit is hit first. To iterate them, you need to pass back the continuation token and will get `has_more` as indicator if there are more results.

**Comment by fundthmcalculus** [comment]
Personally, I would rather not expose the cosmosdb underlying behavior, since that is backend implementation specific. That being said, if the goal of the sane defaults query is to be sane, it should just return everything. I would rather have two call options than 3 with even more permutations. Every time we add options on the SDK side, we multiple the work to implement and support it.

**Comment by tmarkovski** [comment]
I don't think I fully understand what problems that developers might face that we are trying to solve.

If we want to change the default query from `SELECT c.id, c.data, c.type FROM c` and add `OFFSET 0 LIMIT 100`, that's perfectly fine.

If we want to return the entire wallet, even when the size of that wallet is million records, by fetching everything on the server and sending it back to the user, we can, but it's a horrible idea.

If we want to abstract the API to not use SQL at all, let's have that discussion instead.

**Comment by fundthmcalculus** [comment]
> Next question: Do we allow the user to pick offset numbers? Or is the default query only retrieving the first 100 (or fewer) records?

I don't think that's needed, as it'll create confusing method definition, where you can set the `limit`, but also provide fully customized query, which will ignore the `limit` anyway.

Users should have full control over the LIMIT property using the query, as they know best how many records they need returned.

We could also look into adding sort logic in the default query, to ensure that latest 100 records are returned, as an added convenience.

**Comment by sethjback** [comment]
> Another way to query would be not to use pages, but rely on the continuation token functionality, in which case CosmosDB will decide how many results it will return which is 1000 or 1MB result, whichever limit is hit first. To iterate them, you need to pass back the continuation token and will get has_more as indicator if there are more results.

I am somewhat confused by the continuation token in cosmos, because their documentation indicates that some queries (ones that use `group by` or `distinct`) won't work with continuation tokens. So....what if those return over 1mb of results?

I would agree with @fundthmcalculus though that it's probably better to not expose internal storage behavior in the SDK.

> I don't think I fully understand what problems that developers might face that we are trying to solve

This is my fault. In the other discussion about the limit field, I was approaching the problem from a developer convenience / expectation standpoint - It  is nice to have the power to write your own sql, but if I am a developer and just want to "select all" but give it to me in sets of 5 and I want 10 of those, I don't want to have to write my own SQL for that. The default query returns all results, and it is fairly common for APIs to have flags that allow you to paginate the return.

I would definitely be opposed to parsing and modifying user provided SQL - that seems like a non-starter.

> If we want to return the entire wallet, even when the size of that wallet is million records, by fetching everything on the server and sending it back to the user, we can, but it's a horrible idea.

Two thoughts:
1. is it likely there will be a wallet with millions of records?
2. if we went the pagination route we would need to have sane defaults (e.g. 10 items per page, 10 pages by default).

> If we want to abstract the API to not use SQL at all, let's have that discussion instead.

This is me wanting to have my cake and eat it too. It is super powerful to be able to craft your own queries when looking through stuff in your wallet. That seems like a really cool feature. _However_, I will likely NOT need that 90% of the time (an assumption). It seems more likely that I would want a list of all my wallet data to scroll through, and want to be able to paginate it. And for that, my first instinct is going to be to look for API flags, not figure out how to craft a SQL query.

I could of course be completely off base here - maybe the path forward is just lots of examples and documentation so devs can copy / past common patterns.

**Comment by fundthmcalculus** [comment]
> I would definitely be opposed to parsing and modifying user provided SQL - that seems like a non-starter.

👏 this is also disingenuous to the user as it creates side effects.

> and just want to "select all" but give it to me in sets of 5 and I want 10 of those, I don't want to have to write my own SQL for that. 

I would generally agree. Maybe we add an SDK function called `SelectAllPaginated(int offset, int page_size)` and leave the query function alone? This would make it more clear to the user what they have as a convenience function, and we can handle the implementation specifics of paging all the results.

> maybe the path forward is just lots of examples and documentation so devs can copy / past common patterns.

I don't want to have a lot of extraneous functions `SelectFirstTenWalletItems()`, `SelectAllCredentials()`, etc in the SDKs, because that would be confusing. On the other hand, having a set of **tested** SQL queries that are injected right into the docs would be super slick. This would allow us to have our cake and eat it too. We don't have to create a bunch of functions in every language, and our customers don't have to think about writing common SQL queries. Bonus is that it gives us more examples to show what can (and can't) be queried in SQL.

**Comment by geel9** [comment]
I think instead of adding additional special parameters, we could split the search API into two forms:

- SQL query only. Add limits and pagination to your queries if you want said features.
- Structured object which is a query builder lite. This is transformed into a SQL query (by the SDK or by the server) and executed using the pipeline from the first option.


In the second option, you'd pass in, say, a `WalletSearchParams` object, which might have fields like:

- Paging params (`skip/offset`, `limit`)
- `issuer`
- `issueDate`
- `type`
- etc.

These could all be simple equality fields (you'd just set `type` to, say, `"GymMembership"`) or we could go a bit advanced and have each one be a `SearchPredicate`, which could specify a value and an operation (to perform, say, "type in [{type1}, {type2}]" or "issueDate >= {foo}")

This would also probably have a fixed return structure -- that is, they couldn't specify which fields they're selecting or their names. If they want more advanced, they can write a SQL query.

Basically, we can build a query for you from a few of the most-useful components (turning it into a more standard-looking search endpoint), and we entirely control the SQL output of that, _or_ you can just write a SQL query.

**Comment by fundthmcalculus** [comment]
@geel9 I appreciate the idea, but I'm concerned about the complexity of creating a structured query builder. That's a lot of code to write across languages, with a questionable amount of return. What might be more useful is to create a free website/service which does this ?on the docs site? so that the user can take the created SQL query and inject it into their code. I would rather give the user the power and flexibility from the SDK, and provide some of the auxiliary stuff via a separate service. (DRY).

@michaeldboyd @tmarkovski didn't somebody create a simple example of such a structured query builder at one point?

**Comment by tmarkovski** [comment]
Yeah, we want to keep it simple for maintenance. The wallet data can be random, it's not just VCs, so a solution would have to be more general. Exposing the SQL gives us flexibility without over engineering it. I could see us adding some Graph QL capability at some point, though I'm not sure if GraphQL is simpler to use than SQL

**Comment by geel9** [comment]
@fundthmcalculus To clarify, I was thinking no more complicated than a flat struct with scalar values, from which a _very_ simple SQL query would be generated on the server side (meaning minimal SDK work). The additional custom predicate stuff would be nice, but add language complexity; really was just spitballing there.

That said, I agree that the non-uniformity of the data stored makes this pretty much a no-go.

**Comment by fundthmcalculus** [comment]
@geel9 that makes more sense. I still think that might be something better put into the docs site than into the SDK itself.

**Comment by fundthmcalculus** [comment]
@tmarkovski thoughts on if we should handle the `.has_more` and `continuationtoken` inside the SDK ourselves? Relevant bug report: https://github.com/trinsic-id/sdk/issues/583

I lean the direction of handling it on the sane defaults call, which means we should handle it on both. That way, the user isn't tied to our backend implementation.

**Comment by tmarkovski** [comment]
> @tmarkovski thoughts on if we should handle the `.has_more` and `continuationtoken` inside the SDK ourselves? Relevant bug report: #583
> 
> I lean the direction of handling it on the sane defaults call, which means we should handle it on both. That way, the user isn't tied to our backend implementation.

Only if we utilize grpc server streaming, but even then it's not desirable. Otherwise, `continuation token` logic must work as intended by the DB. We shouldn't pull all records everytime someone does a search. It'll both eat up the DB throughput, kill the UI memory and create horrible scenarios with super large data sets.

**Comment by fundthmcalculus** [comment]
> > @tmarkovski thoughts on if we should handle the `.has_more` and `continuationtoken` inside the SDK ourselves? Relevant bug report: #583
> > I lean the direction of handling it on the sane defaults call, which means we should handle it on both. That way, the user isn't tied to our backend implementation.
> 
> Only if we utilize grpc server streaming, but even then it's not desirable. Otherwise, `continuation token` logic must work as intended by the DB. We shouldn't pull all records everytime someone does a search. It'll both eat up the DB throughput, kill the UI memory and create horrible scenarios with super large data sets.

In that event, let's limit the "sane default query" to `LIMIT 100`.

**Comment by fundthmcalculus** [comment]
https://github.com/trinsic-id/sdk-swift/pull/13

**Comment by tmarkovski** [comment]
Just a note that will still not limit the records to 100 if the result set exceeds the max payload.

**Comment by geel9** [comment]
Do you mean that it might return, for example, 50 items out of an 85-item result set -- or that it may return *more than 100* items at once?

**Comment by fundthmcalculus** [comment]
> Do you mean that it might return, for example, 50 items out of an 85-item result set -- or that it may return _more than 100_ items at once?

50 out of 85, not more than 100. Max it would return is 1 MB of data, or 1000 records, whichever is less. That's what the `continuation_token` and `has_more` parameters are.
